src/MobxPromise.ts

```diff
@@ -153,38 +152,52 @@ export class MobxPromiseImpl<R>
 				if (error)
 					return error;
 
-		return this.internalError;
+		return this.internalError.value;
 	}
 
 	/**
 	 * This lets mobx determine when to call this.invoke(),
 	 * taking advantage of caching based on observable property access tracking.
 	 */
-	@computed private get latestInvokeId()
+	@computed private get invokeResult()
 	{
-		window.clearTimeout(this._latestInvokeId);
-		let promise = this.invoke();
-		let invokeId:number = window.setTimeout(() => this.setPending(invokeId, promise));
-		return this._latestInvokeId = invokeId;
+		let syncResult = {
+			isSet: false,
+			value: (undefined as R|undefined)
+		};
+
+		let promise = this.invoke((result:R)=>{
+			syncResult.isSet = true;
+			syncResult.value = result;
+		});
+
+		// override promise value, and set synchronously, if theres a synchronous resolution
+		if (syncResult.isSet) {
+			this.setComplete(this._asyncInvokeCount, syncResult.value!);
+			return 0;
+		} else {
+			this._asyncInvokeCount += 1;
+			this.setPending(this._asyncInvokeCount, promise);
+			return 0;
+		}
```

Comment:
At this point it's not exactly well suited to be a @computed. But it wouldn't work as an autorun either because that would break the system where an unreferenced mxp goes stale. So I don't know. I think this is the best we can do.